packages/@aws-cdk/cloudformation-diff/lib/diff/template-and-changeset-diff-merger.ts

```diff
@@ -0,0 +1,211 @@
+        props[propertyName] = value.afterValue;
+      }
+    } else {
+      for (const [propertyName, value] of Object.entries(resourceInfoFromChangeset.properties ?? {})) {
+        props[propertyName] = value.beforeValue;
+      }
+    }
+
+    return {
+      Type: resourceInfoFromChangeset.resourceType ?? TemplateAndChangeSetDiffMerger.UNKNOWN_RESOURCE_TYPE,
+      Properties: props,
+    };
+  }
+
+  static determineChangeSetReplacementMode(propertyChange: ChangeSetResourceChangeDetail): types.ChangeSetReplacementMode {
+    if (propertyChange.Target?.RequiresRecreation === undefined) {
+      // We can't determine if the resource will be replaced or not. That's what conditionally means.
+      return 'Conditionally';
+    }
+
+    if (propertyChange.Target.RequiresRecreation === 'Always') {
+      switch (propertyChange.Evaluation) {
+        case 'Static':
+          return 'Always';
+        case 'Dynamic':
+          // If Evaluation is 'Dynamic', then this may cause replacement, or it may not.
+          // see 'Replacement': https://docs.aws.amazon.com/AWSCloudFormation/latest/APIReference/API_ResourceChange.html
+          return 'Conditionally';
+      }
+    }
+
+    return propertyChange.Target.RequiresRecreation as types.ChangeSetReplacementMode;
+  }
+
+  changeSet: DescribeChangeSetOutput | undefined;
+  changeSetResources: types.ChangeSetResources;
+
+  constructor(
+    args: {
+      changeSet: DescribeChangeSetOutput;
+      changeSetResources?: types.ChangeSetResources; // used for testing -- otherwise, don't populate
+    },
+  ) {
+    this.changeSet = args.changeSet;
+    this.changeSetResources = args.changeSetResources ?? this.createChangeSetResources(this.changeSet);
+  }
+
+  /**
+   * use information from the changeset to populate details that are missing in the templateDiff
+   */
+  createChangeSetResources(changeSet: DescribeChangeSetOutput): types.ChangeSetResources {
+    const changeSetResources: types.ChangeSetResources = {};
+    for (const resourceChange of changeSet.Changes ?? []) {
+      if (resourceChange.ResourceChange?.LogicalResourceId === undefined) {
+        continue;
+      }
+
+      const propertiesReplaced: types.ChangeSetProperties = {};
+      for (const propertyChange of resourceChange.ResourceChange.Details ?? []) {
+        if (propertyChange.Target?.Attribute === 'Properties' && propertyChange.Target.Name) {
+          propertiesReplaced[propertyChange.Target.Name] = {
+            changeSetReplacementMode: TemplateAndChangeSetDiffMerger.determineChangeSetReplacementMode(propertyChange),
+            beforeValue: _maybeJsonParse(propertyChange.Target.BeforeValue),
+            afterValue: _maybeJsonParse(propertyChange.Target.AfterValue),
+          };
+        }
+      }
+
+      changeSetResources[resourceChange.ResourceChange.LogicalResourceId] = {
+        resourceWasReplaced: resourceChange.ResourceChange.Replacement === 'True',
```

Comment:
Are we sure `ResourceChange` will always exist? I'm seeing that the old code always had `resourceChange.ResourceChange?.<property>`.

Reply:
I guess line 80 would be our check for this?

Reply:
yeah, I'm trying my best to be super paranoid in this PR about running into anything that's undefined... part of me wants to wrap the entire changeset operation in try-catch and just continue with template diff if the changeset operations fail... but that would probably not be good in that it would mask many issues.